[IR] Introduce `mem.cache_hint` metadata for composable memory cache control hints (#181612)
llvm/llvm-project: llvm/docs/LangRef.rst

uses of the stored-to memory location after it has been overwritten with a
different value.

'``mem.cache_hint``' Metadata
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``!mem.cache_hint`` metadata may be attached to any instruction that reads
or writes memory. It provides target-specific cache control hints for the
memory operation. This metadata is a performance hint: dropping or ignoring it
must not change the observable behavior of the program.

The value of ``!mem.cache_hint`` is a single metadata node containing a flat
list of ``(operand_no, hint_node)`` pairs. Each ``operand_no`` is an ``i32``
constant that is the raw IR operand index (for regular instructions) or the

> argument index (for intrinsic calls) of a pointer-typed operand. Each

Reviewer: Was not immediately clear to me, how about the following? (Also shouldn't it be quite implicit the operand index is an unsigned?)
```suggestion
list of ``(operand_no, hint_node)`` pairs. For non-call instructions, ``operand_no`` is the IR operand index. For intrinsic calls, ``operand_no`` is the call argument index. Each
```

``hint_node`` is a metadata node containing target-prefixed key/value string
pairs.

The ``!mem.cache_hint`` node must contain an even number of entries, alternating
``i32`` operand numbers and metadata nodes. Operand numbers must be unique within
a ``!mem.cache_hint`` node and must refer to a pointer-typed operand. Keys within
a single hint node must also be unique.

For a ``load``, the pointer is operand 0. For a ``store``, the value is
operand 0 and the pointer is operand 1. For intrinsic calls such as
``llvm.memcpy``, operand numbers correspond to argument positions: e.g.,
destination is argument 0 and source is argument 1.

The hint node keys are prefixed with a target identifier (e.g., ``nvvm.``) and
their interpretation is entirely target-dependent. The IR verifier enforces only
the structural rules above; validation of target-specific keys and values is
performed by the corresponding backend. Unsupported properties may be silently
ignored during code generation.

The following examples use ``nvvm.`` prefixed keys for NVIDIA GPU targets.
Other targets may define their own prefixed keys.

> Example: load with cache hints (NVIDIA GPU):

Reviewer: Nit: I might suggest dropping NVIDIA GPU in the examples list, we have already mentioned above "The following examples... for NVIDIA GPU", which I think should suffice (and might as well move "Other targets" part after the examples).

uses of the stored-to memory location after it has been overwritten with a
different value.

'``mem.cache_hint``' Metadata
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``!mem.cache_hint`` metadata may be attached to any instruction that reads
or writes memory. It provides target-specific cache control hints for the
memory operation. This metadata is a performance hint: dropping or ignoring it
must not change the observable behavior of the program.

The value of ``!mem.cache_hint`` is a single metadata node containing a flat
list of ``(operand_no, hint_node)`` pairs. For non-call instructions, ``operand_no``

> is the IR operand index. For intrinsic calls, ``operand_no`` is the call argument index.

Reviewer 2: Text needs updating? Wasn't operand_no removed from the first draft? It's not in the examples below

Reviewer 3: You're right, I do think the wording is a bit confusing.

uses of the stored-to memory location after it has been overwritten with a
different value.

'``mem.cache_hint``' Metadata
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``!mem.cache_hint`` metadata may be attached to any instruction that reads
or writes memory. It provides target-specific cache control hints for the
memory operation. This metadata is a performance hint: dropping or ignoring it
must not change the observable behavior of the program.

The ``!mem.cache_hint`` node must contain an even number of entries, alternating between
``i32`` operand numbers and metadata hint nodes.

Operand numbers refer to the pointer operand that carries the cache hint specified by the following hint node.
For non-call instructions, the operand number is the IR operand index.
For intrinsic calls, the operand number is the call argument index.
Operand numbers must be unique within
a ``!mem.cache_hint`` node and must refer to a pointer-typed operand.

> For exmaple, for a ``load``, the pointer is operand 0. For a ``store``, the value is

Reviewer 4: ```suggestion
For example, for a ``load``, the pointer is operand 0. For a ``store``, the value is
```

uses of the stored-to memory location after it has been overwritten with a
different value.

'``mem.cache_hint``' Metadata
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``!mem.cache_hint`` metadata may be attached to any instruction that reads

> or writes memory. It provides target-specific cache control hints for the

Reviewer: Considering that the allowed calls that may read/write memory are now only recognized intrinsics, I believe this could be made slightly more precise:

```suggestion
The ``!mem.cache_hint`` metadata may be attached to any instruction that reads
or writes memory, except non-intrinsic calls. It provides target-specific cache control hints for the
```

As one may wonder why we do this, I'd personally include the rationale too (i.e., as already suggested above by @nikic, any generic function call is rejected because their memory behavior depends on attributes that may change independently), but not strong on this.

memory operation. This metadata is a performance hint: dropping or ignoring it

> must not change the observable behavior of the program.

Reviewer: The wording here should be already fine, though I wonder if we should be nitpicking, along what @nikic and @RalfJung mentioned in the RFC, i.e., that the cache hints are strictly outside of the LLVM memory model (one could accidentally reason by analogy w/ `!nontemporal`, which is a hint too but has ordering implications):

```suggestion
memory operation. This metadata is purely a performance hint: dropping or ignoring it
must not change the observable behavior of the program (particularly, cache hints do not affect memory model semantics).
```

uses of the stored-to memory location after it has been overwritten with a
different value.

'``mem.cache_hint``' Metadata
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``!mem.cache_hint`` metadata may be attached to any instruction that reads
or writes memory, except non-intrinsic calls. We do not allow cache hints

> on function calls because their memory behavior depends on attributes that may change independently, and may therefore invalidate the IR.

Reviewer: Nit: let's favor passive construction in LangRef when possible.
```suggestion
or writes memory, except non-intrinsic calls. Cache hints are not
permitted on function calls because their memory behavior
depends on attributes, which may change independently.
```

The ``!mem.cache_hint`` metadata provides target-specific cache control hints for the
memory operation. This metadata is purely a performance hint: dropping or ignoring it
must not change the observable behavior of the program (particularly, cache hints do not affect memory model semantics).

The ``!mem.cache_hint`` node must contain an even number of entries, alternating between
``i32`` operand numbers and metadata hint nodes.

> Operand numbers refer to the pointer operand that carries the cache hint specified by the following hint node.

Reviewer: ```suggestion
Each operand number identifies the pointer operand to which the following hint node applies.
```

uses of the stored-to memory location after it has been overwritten with a
different value.

'``mem.cache_hint``' Metadata

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 5: Please either keep to 80 characters or do per-sentence line breaks. Currently there is a weird mix of everything going on in here.
